packages/hub/src/lib/cache-management.ts

```diff
@@ -0,0 +1,257 @@
+	const cachedRevisions: Set<CachedRevisionInfo> = new Set();
+	const blobStats: Map<string, Stats> = new Map(); // Store blob stats
+
+	const snapshotDirs = await readdir(snapshotsPath);
+	for (const dir of snapshotDirs) {
+		if (FILES_TO_IGNORE.includes(dir)) continue; // Ignore unwanted files
+
+		const revisionPath = join(snapshotsPath, dir);
+		const revisionStat = await stat(revisionPath);
+		if (!revisionStat.isDirectory()) {
+			throw new Error(`Snapshots folder corrupted. Found a file: ${revisionPath}`);
+		}
+
+		const cachedFiles: Set<CachedFileInfo> = new Set();
+		await scanSnapshotDir(revisionPath, cachedFiles, blobStats);
+
+		const revisionLastModified =
+			cachedFiles.size > 0 ? Math.max(...[...cachedFiles].map((file) => file.blobLastModified)) : revisionStat.mtimeMs;
+
+		cachedRevisions.add({
+			commitHash: dir,
+			files: cachedFiles,
+			refs: refsByHash.get(dir) || new Set(),
+			sizeOnDisk: [...cachedFiles].reduce((sum, file) => sum + file.sizeOnDisk, 0),
+			snapshotPath: revisionPath,
+			lastModified: revisionLastModified,
+		});
+
+		refsByHash.delete(dir);
+	}
+
+	// Verify that all refs refer to a valid revision
+	if (refsByHash.size > 0) {
+		throw new Error(
+			`Reference(s) refer to missing commit hashes: ${JSON.stringify(Object.fromEntries(refsByHash))} (${repoPath})`
+		);
+	}
+
+	const repoStats = await stat(repoPath);
+	const repoLastAccessed =
+		blobStats.size > 0 ? Math.max(...[...blobStats.values()].map((stat) => stat.atimeMs)) : repoStats.atimeMs;
+
+	const repoLastModified =
+		blobStats.size > 0 ? Math.max(...[...blobStats.values()].map((stat) => stat.mtimeMs)) : repoStats.mtimeMs;
+
+	// Return the constructed CachedRepoInfo object
+	return {
+		repoId: repoId,
+		repoType: repoType,
+		repoPath: repoPath,
+		nbFiles: blobStats.size,
+		revisions: cachedRevisions,
+		sizeOnDisk: [...blobStats.values()].reduce((sum, stat) => sum + stat.size, 0),
+		lastAccessed: repoLastAccessed,
+		lastModified: repoLastModified,
+	};
+}
+
+export async function scanRefsDir(refsPath: string, refsByHash: Map<string, Set<string>>): Promise<void> {
+	const refFiles = await readdir(refsPath, { withFileTypes: true });
+	for (const refFile of refFiles) {
+		const refFilePath = join(refsPath, refFile.name);
+		if (refFile.isDirectory()) continue; // Skip directories
+
+		const commitHash = await readFile(refFilePath, "utf-8");
+		const refName = refFile.name;
+		if (!refsByHash.has(commitHash)) {
+			refsByHash.set(commitHash, new Set());
+		}
+		refsByHash.get(commitHash)?.add(refName);
+	}
+}
+
+export async function scanSnapshotDir(
+	revisionPath: string,
+	cachedFiles: Set<CachedFileInfo>,
+	blobStats: Map<string, Stats>
+): Promise<void> {
+	const files = await readdir(revisionPath, { withFileTypes: true });
+	for (const file of files) {
+		if (file.isDirectory()) continue; // Skip directories
+
+		const filePath = join(revisionPath, file.name);
+		const blobPath = await realpath(filePath);
```

Comment:
Is `fs.realpath`  different from `path.resolve` ?

First time I see `fs.realpath`, I'd rather use `path.resolve` if it's the same

Reply:
> The fs.realPath() method is used to compute the canonical pathname of the given path. It does so by resolving the ., .. and the symbolic links in the path.[^1]

The part we are interested is **the symbolic links in the path**. 

Giving a symlinks to a path.resolve would not have the expected behaviour here.

Here is a short demo showing the difference between `fs.realpath` and `path.resolve`.

```
$: echo "Hello world" > example.txt
$: ln -s example.txt symlink.txt
$: ls
example.txt symlink.txt
$: node
> const fs = require("node:fs/promises")
> await fs.realpath("symlink.txt")
'/tmp/example.txt'
> const path = require("node:path")
> path.resolve("symlink.txt")
'/tmp/symlink.txt'
```


[^1]: https://www.geeksforgeeks.org/node-js-fs-realpath-method/